Leanplum-SDK/Classes/Managers/Networking/LPRequestSender.m

```diff
@@ -66,53 +65,67 @@ - (id)init
     if (self) {
         if (_engine == nil) {
             if (!_requestHeaders) {
-                _requestHeaders = [LPRequestSender createHeaders];
+                _requestHeaders = [LPNetworkEngine createHeaders];
             }
             _engine = [LPNetworkFactory engineWithHostName:[LPConstantsState sharedState].apiHostName
                                         customHeaderFields:_requestHeaders];
         }
+        [LPRequestSenderTimer start];
         _countAggregator = [LPCountAggregator sharedAggregator];
     }
     return self;
 }
 
 - (void)send:(LPRequest *)request
 {
-
-    [self sendEventually:request sync:NO];
-    if ([LPConstantsState sharedState].isDevelopmentModeEnabled) {
-        NSTimeInterval currentTime = [NSDate timeIntervalSinceReferenceDate];
-        NSTimeInterval delay;
-        if (!self.lastSentTime || currentTime - self.lastSentTime > LP_REQUEST_DEVELOPMENT_MAX_DELAY) {
-            delay = LP_REQUEST_DEVELOPMENT_MIN_DELAY;
-        } else {
-            delay = (self.lastSentTime + LP_REQUEST_DEVELOPMENT_MAX_DELAY) - currentTime;
+    [self saveRequest:request sync:request.sendSynchronously];
+    if ([LPConstantsState sharedState].isDevelopmentModeEnabled || request.requestType == Immediate) {
+        if ([self validateConfigFor:request]) {
+            if (request.datas != nil) {
+                [self sendNow:request withDatas:request.datas];
+            } else {
+                [self sendNow:request];
+            }
         }
-        [self performSelector:@selector(sendIfConnected:) withObject:request afterDelay:delay];
     }
-    [self.countAggregator incrementCount:@"send_request_lp"];
 }
 
-- (void)sendNow:(LPRequest *)request sync:(BOOL)sync
+- (BOOL)validateConfigFor:(LPRequest *)request
 {
-    RETURN_IF_TEST_MODE;
-
     if (![LPAPIConfig sharedConfig].appId) {
         LPLog(LPError, @"Cannot send request. appId is not set");
-        return;
+        return false;
     }
+    
     if (![LPAPIConfig sharedConfig].accessKey) {
         LPLog(LPError, @"Cannot send request. accessKey is not set");
-        return;
+        return false;
+    }
+    
+    if (![[Leanplum_Reachability reachabilityForInternetConnection] isReachable])
+    {
+        LPLog(LPError, @"Device is offline, will try sending requests again later.");
+        if (request.errorBlock) {
+            request.errorBlock([NSError errorWithDomain:@"Leanplum" code:1
+                                               userInfo:@{NSLocalizedDescriptionKey: @"Device is offline"}]);
+        }
+
+        return false;
     }
+    
+    return true;
+}
+
+- (void)sendNow:(LPRequest *)request sync:(BOOL)sync
```

Comment:
Please get rid off `sync` flag from all methods and always execute on `LPOperationQueue`